central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -202,45 +293,144 @@ func (ds *datastoreImpl) GetBatch(ctx context.Context, ids []string) ([]*storage
+		return fmt.Errorf("collections must have non-empty, unique `name` values (%s)", collection.GetName())
+	}
+
+	// add to graph to detect any cycles, this also sets the `id` field
+	graph, err := ds.addCollectionToGraph(collection)
 	if err != nil {
 		return err
 	}
 
+	// if this is a dryrun, we don't want to add to storage or make changes to objects
+	if dryrun {
+		collection.Id = ""
+		return nil
+	}
+
+	// add to storage
 	err = ds.storage.Upsert(ctx, collection)
 	if err != nil {
-		// if we fail to upsert, update the graph
-		deleteErr := ds.deleteCollectionFromGraph(collection.GetId())
-		if deleteErr != nil {
-			log.Errorf("Failed to remove collection from internal state object (%v)", deleteErr)
-		}
 		return err
 	}
-	return err
+
+	// we've succeeded, now set all the values
+	ds.names.Add(collection.GetName())
+	ds.graph = graph
+	return nil
+}
+
+func (ds *datastoreImpl) AddCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+	return ds.addCollectionWorkflow(ctx, collection, false)
 }
 
 func (ds *datastoreImpl) DryRunAddCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+
 	// check for access since dryrun flow doesn't actually hit the postgres layer
 	if ok, err := workflowSAC.WriteAllowed(ctx); err != nil || !ok {
 		return err
 	}
 
-	return ds.dryRunAddCollectionInGraph(collection)
+	return ds.addCollectionWorkflow(ctx, collection, true)
+}
+
+func (ds *datastoreImpl) updateCollectionWorkflow(ctx context.Context, collection *storage.ResourceCollection, dryrun bool) error {
+
+	// sanity checks
+	if err := verifyCollectionObjectNotEmpty(collection); err != nil {
+		return err
+	}
+	if collection.GetId() == "" {
+		return errors.New("update must be called on an existing collection")
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	// resolve object to check if the name was changed
+	obj, ok, err := ds.storage.Get(ctx, collection.GetId())
+	if err != nil || !ok {
+		return errors.Wrap(err, "failed to resolve collection being updated")
+	}
+	if obj.GetName() != collection.GetName() && ds.names.Contains(collection.GetName()) {
+		return fmt.Errorf("collection name already in use (%s)", collection.GetName())
+	}
+
+	// update graph first to detect cycles
+	graph, err := ds.updateCollectionInGraph(collection)
+	if err != nil {
+		return err
+	}
+
+	// if this is a dryrun we don't want to make changes to the datastore or tracking objects
+	if dryrun {
+		return nil
+	}
+
+	// update datastore
+	err = ds.storage.Upsert(ctx, collection)
+	if err != nil {
+		return err
+	}
+
+	// success, we now update objects
+	if obj.GetName() != collection.GetName() {
+		ds.names.Remove(obj.GetName())
+		ds.names.Add(collection.GetName())
+	}
+	ds.graph = graph
+	return nil
+}
+
+func (ds *datastoreImpl) UpdateCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+	return ds.updateCollectionWorkflow(ctx, collection, false)
```

Comment:
sac check

Reply:
Same comment as on the AddCollection function